### ISSUE TITLE
Wayland support?

### ISSUE THREAD

**Issue by mwhipple** [9fans/plan9port#130]
When trying to run acme on a Linux amd64 box with Wayland (no XWayland) I'm getting:
```
acme: can't open display: muxrpc: unexpected eof
```
Is this due to missing X stuff? Is Wayland support planned?

If the issue is missing support rather than something solely on my end, I'd be interested in contributing in whatever way I can.

**Comment by osvein** [comment]
Right now the only devdraw backends are for Xlib, Carbon and Cocoa.

devdraw is a bit messy IMO, but if you're experienced with Wayland it shouldn't be too difficult to port. I've been working on [an SDL 2.0 backend for it](https://github.com/osvein/devdraw-sdl) without much prior experience with either libdraw or SDL, but it's looking good so far. I'm probably not going to be testing it on Wayland, but SDL 2.0 does have Wayland support IIRC.

Though XWayland is probably the easiest way out.

**Comment by sirnewton01** [comment]
There is now an experimental wayland support for drawterm here.
https://github.com/driusan/drawterm

It seems that it should be portable to devdraw somehow. I can't seem to figure out how devdraw abstracts Cocoa/Carbon/X11 though.

**Comment by eaburns** [comment]
I know this is an old issue, but I spent the weekend learning Wayland enough to add a devdraw driver for it. The code is here: https://github.com/eaburns/plan9port/tree/wayland. It's still not well tested, but seems to work OK for my setup.

**Comment by eaburns** [comment]
@rsc any interest in me sending a pull request to add wayland support (see my branch linked in the comment above)? Or is it best to leave it as a fork?

**Comment by dancrossnyc** [comment]
Go ahead and send the PR; we'll have a look.

**Comment by eaburns** [comment]
I see. If we can get an implementation with wp_pointer_warp_v1, then I can test it out. I'm not sure how new it is; it's possible that it didn't exist when I was creating this backend.

(In general, I'm sort of disappointed with how inconsistent Wayland implementations are with all these extensions, but I guess that's just how things are when you are using non-standard features.)

**Comment by yrk-lab** [comment]
#696 implements this.

**Comment by lobre** [comment]
There is this project https://github.com/blakemcbride/Acme (which seems to be vibe-coded) which implemented an SDL3 backend for devdraw, which naturally brings Wayland.

This could be another direction to tackle Wayland support.

**Comment by dancrossnyc** [comment]
Yes. A PR was sent to incorporate that, but the vibes did not meet the quality bar we expect, so it was rejected.

**Comment by lobre** [comment]
Effectively, I did not see that as the PR was not linked here. Here it is: https://github.com/9fans/plan9port/pull/763. Thanks for answering.

I am not advocating taking contributions from LLM-generated code at all (I am also rather against this practice, especially on well-crafted open-source projects such as plan9port). I was just curious about this angle of implementing devdraw with SDL3, which gives Wayland for free. So I wanted to reference it here. But it seems you are already aware.

I understand that everyone's time is precious, and Wayland is a big beast. I have hope that someone (me included) would make time someday to tackle that challenge once and for all, as Wayland is more and more common on Linux desktops.

I also wanted to note that SDL3 seems to come with client-side decoration included for compositors such as Mutter, which forces them (which was one of the challenges met in other PRs from @eaburns).
